File: ios/AiStroykaWorker/AiStroykaWorker/Services/WorkerAPI.swift
//
//  WorkerAPI.swift
//  AiStroykaWorker
//

import Foundation
import Shared

/// High-level API for AiStroyka Worker endpoints. All writes use idempotency keys.
enum WorkerAPI {
    
    static func config() async throws -> ConfigResponse {
        try await APIClient.shared.request(path: "config")
    }
    
    static func projects() async throws -> [ProjectDTO] {
        let r: ProjectsResponse = try await APIClient.shared.request(path: "projects")
        return r.data ?? []
    }

    /// POST /api/v1/devices/register — register APNS token for push. Token must never be logged or shown in UI.
    static func registerDevice(pushToken: String) async throws {
        struct Body: Encodable {
            let deviceId: String
            let platform: String
            let token: String
            enum CodingKeys: String, CodingKey {
                case deviceId = "device_id"
                case platform
                case token
            }
        }
        let body = Body(deviceId: DeviceContext.deviceId, platform: "ios", token: pushToken)
        let _: RegisterDeviceResponse = try await APIClient.shared.request(
            path: "devices/register",
            method: "POST",
            body: body,
            idempotencyKey: DeviceContext.idempotencyKeyDeviceRegister(pushToken: pushToken)
        )
    }

    /// POST /api/v1/devices/unregister — remove this device row server-side (call before sign-out when session exists).
    static func unregisterDevice() async throws {
        struct Body: Encodable {
            let deviceId: String
            enum CodingKeys: String, CodingKey { case deviceId = "device_id" }
        }
        let _: RegisterDeviceResponse = try await APIClient.shared.request(
            path: "devices/unregister",
            method: "POST",
            body: Body(deviceId: DeviceContext.deviceId),
            idempotencyKey: DeviceContext.idempotencyKeyDeviceUnregister()
        )
    }

    /// GET /api/v1/worker/tasks/today?project_id= (optional)
    static func tasksToday(projectId: String?) async throws -> [TaskDTO] {
        var path = "worker/tasks/today"
        if let id = projectId, !id.isEmpty {
            path += "?project_id=\(id.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? id)"
        }
        let r: TasksTodayResponse = try await APIClient.shared.request(path: path)
        return r.data ?? []
    }
    
    static func startDay(idempotencyKey: String) async throws {
        try await APIClient.shared.requestVoid(
            path: "worker/day/start",
            method: "POST",
            body: EmptyBody(),
            idempotencyKey: idempotencyKey
        )
    }
    
    static func endDay(idempotencyKey: String) async throws {
        try await APIClient.shared.requestVoid(
            path: "worker/day/end",
            method: "POST",
            body: EmptyBody(),
            idempotencyKey: idempotencyKey
        )
    }
    
    static func createReport(dayId: String?, taskId: String?, idempotencyKey: String) async throws -> String {
        struct Body: Encodable {
            let dayId: String?
            let taskId: String?
            enum CodingKeys: String, CodingKey { case dayId = "day_id"; case taskId = "task_id" }
        }
        let r: ReportCreateResponse = try await APIClient.shared.request(
            path: "worker/report/create",
            method: "POST",
            body: Body(dayId: dayId, taskId: taskId),
            idempotencyKey: idempotencyKey
        )
        guard let id = r.data?.id else { throw APIError(statusCode: nil, code: nil, message: "No report id in response") }
        return id
    }
    
    static func addMedia(reportId: String, uploadSessionId: String, idempotencyKey: String) async throws {
        struct Body: Encodable {
            let reportId: String
            let uploadSessionId: String
            enum CodingKeys: String, CodingKey {
                case reportId = "report_id"
                case uploadSessionId = "upload_session_id"
            }
        }
        try await APIClient.shared.requestVoid(
            path: "worker/report/add-media",
            method: "POST",
            body: Body(reportId: reportId, uploadSessionId: uploadSessionId),
            idempotencyKey: idempotencyKey
        )
    }
    
    static func submitReport(reportId: String, taskId: String?, workerNote: String?, idempotencyKey: String) async throws {
        struct Body: Encodable {
            let reportId: String
            let taskId: String?
            let workerNote: String?
            enum CodingKeys: String, CodingKey {
                case reportId = "report_id"
                case taskId = "task_id"
                case workerNote = "worker_note"
            }
        }
        let trimmed = workerNote?.trimmingCharacters(in: .whitespacesAndNewlines)
        let notePayload: String? = {
            guard let t = trimmed, !t.isEmpty else { return nil }
            return String(t.prefix(2000))
        }()
        let body = Body(reportId: reportId, taskId: taskId, workerNote: notePayload)
        try await APIClient.shared.requestVoid(
            path: "worker/report/submit",
            method: "POST",
            body: body,
            idempotencyKey: idempotencyKey
        )
    }
    
    static func createUploadSession(purpose: String, idempotencyKey: String) async throws -> (id: String, uploadPath: String) {
        struct Body: Encodable {
            let purpose: String
        }
        let r: UploadSessionResponse = try await APIClient.shared.request(
            path: "media/upload-sessions",
            method: "POST",
            body: Body(purpose: purpose),
            idempotencyKey: idempotencyKey
        )
        guard let data = r.data else { throw APIError(statusCode: nil, code: nil, message: "No session data") }
        let path = data.uploadPath ?? "media/\(data.id)"
        return (data.id, path)
    }
    
    static func finalizeUploadSession(sessionId: String, objectPath: String, mimeType: String?, sizeBytes: Int?, idempotencyKey: String) async throws {
        struct Body: Encodable {
            let objectPath: String
            let mimeType: String?
            let sizeBytes: Int?
            enum CodingKeys: String, CodingKey {
                case objectPath = "object_path"
                case mimeType = "mime_type"
                case sizeBytes = "size_bytes"
            }
        }
        try await APIClient.shared.requestVoid(
            path: "media/upload-sessions/\(sessionId)/finalize",
            method: "POST",
            body: Body(objectPath: objectPath, mimeType: mimeType, sizeBytes: sizeBytes),
            idempotencyKey: idempotencyKey
        )
    }

    // MARK: - Sync (Phase 7.2)
    static func syncBootstrap() async throws -> SyncBootstrapResponse {
        try await APIClient.shared.request(path: "sync/bootstrap")
    }

    /// On 409 returns SyncConflictError with body (mustBootstrap, serverCursor). Caller should bootstrap and retry when mustBootstrap.
    static func syncChanges(cursor: Int, limit: Int = 100) async throws -> SyncChangesResponse {
        let (data, code) = try await APIClient.shared.requestDataAndResponse(path: "sync/changes?cursor=\(cursor)&limit=\(limit)")
        if code == 409 {
            if let body = try? JSONDecoder().decode(SyncConflictBody.self, from: data) {
                throw SyncConflictError(body: body)
            }
            throw APIError.from(data: data, response: nil)
        }
        if code >= 400 {
            throw APIError.from(data: data, response: nil)
        }
        return try JSONDecoder().decode(SyncChangesResponse.self, from: data)
    }

    static func syncAck(cursor: Int, idempotencyKey: String) async throws {
        struct Body: Encodable {
            let cursor: Int
        }
        try await APIClient.shared.requestVoid(
            path: "sync/ack",
            method: "POST",
            body: Body(cursor: cursor),
            idempotencyKey: idempotencyKey
        )
    }

    /// GET /api/v1/activation/status — onboarding/checklist progress.
    static func activationStatus() async throws -> WorkerActivationStatusDTO {
        try await APIClient.shared.request(path: "activation/status")
    }

    /// POST /api/v1/help/hints — localized hints for next steps.
    static func helpHints(locale: String, role: String, getStarted: WorkerGetStartedStatusDTO?) async throws -> [WorkerHelpHintDTO] {
        struct Body: Encodable {
            let locale: String
            let role: String
            let getStarted: WorkerGetStartedStatusDTO?
        }
        let r: WorkerHelpHintsResponseDTO = try await APIClient.shared.request(
            path: "help/hints",
            method: "POST",
            body: Body(locale: locale, role: role, getStarted: getStarted)
        )
        return r.hints ?? []
    }

    static func helpAssistant(
        query: String,
        locale: String,
        role: String,
        pathname: String,
        activation: WorkerActivationStatusDTO?
    ) async throws -> WorkerHelpAssistantResponseDTO {
        struct Body: Encodable {
            let query: String
            let locale: String
            let role: String
            let pathname: String
            let getStarted: WorkerGetStartedStatusDTO?
            let projectCount: Int?
            let taskCount: Int?
            let reportCount: Int?
            let hasAiInsight: Bool?
        }
        let body = Body(
            query: query,
            locale: locale,
            role: role,
            pathname: pathname,
            getStarted: activation?.getStarted,
            projectCount: activation?.projectCount,
            taskCount: activation?.taskCount,
            reportCount: activation?.reportCount,
            hasAiInsight: activation?.hasAiInsight
        )
        return try await APIClient.shared.request(path: "help/assistant", method: "POST", body: body)
    }

    static func helpAssistantEvent(type: String, locale: String, role: String, pathname: String) async {
        struct Body: Encodable {
            let type: String
            let locale: String
            let role: String
            let pathname: String
        }
        _ = try? await APIClient.shared.request(
            path: "help/assistant/events",
            method: "POST",
            body: Body(type: type, locale: locale, role: role, pathname: pathname)
        ) as WorkerHelpAssistantEventAckDTO
    }

    /// GET /api/v1/worker/sync — recent tasks/reports/sessions for the signed-in worker (lite).
    static func workerSync() async throws -> [WorkerSyncReportRow] {
        let env: WorkerSyncEnvelope = try await APIClient.shared.request(path: "worker/sync")
        return env.data?.reports ?? []
    }

    /// GET /api/v1/reports/:id — own report detail (manager note visible to worker).
    static func reportDetail(id: String) async throws -> WorkerReportDetailData {
        let env: WorkerReportDetailEnvelope = try await APIClient.shared.request(path: "reports/\(id)")
        guard let data = env.data else {
            throw APIError(statusCode: nil, code: nil, message: "No report data")
        }
        return data
    }
}

private struct EmptyBody: Encodable {}

struct WorkerActivationStatusDTO: Decodable {
    let projectCount: Int?
    let hasInvited: Bool?
    let taskCount: Int?
    let reportCount: Int?
    let hasAiInsight: Bool?
    let showOnboarding: Bool?
    let getStarted: WorkerGetStartedStatusDTO?
}

struct WorkerGetStartedStatusDTO: Codable {
    let createProject: Bool?
    let inviteTeam: Bool?
    let addTask: Bool?
    let uploadReport: Bool?
    let viewAi: Bool?
}

struct WorkerHelpHintDTO: Decodable {
    let step: String
    let title: String
    let reason: String
    let action: String
    let href: String
}

struct WorkerHelpHintsResponseDTO: Decodable {
    let hints: [WorkerHelpHintDTO]?
}

struct WorkerHelpAssistantRiskSignalDTO: Decodable {
    let id: String
    let title: String
    let detail: String
    let severity: String
    let href: String
}

struct WorkerHelpAssistantResponseDTO: Decodable {
    let summary: String
    let answer: String
    let riskSignals: [WorkerHelpAssistantRiskSignalDTO]?
    let confidence: Int?
}

struct WorkerHelpAssistantEventAckDTO: Decodable {
    let ok: Bool?
}

struct WorkerSyncReportRow: Decodable, Identifiable {
    let id: String
    let status: String
    let createdAt: String
    let submittedAt: String?
}

private struct WorkerSyncEnvelope: Decodable {
    let data: WorkerSyncPayload?
}

private struct WorkerSyncPayload: Decodable {
    let reports: [WorkerSyncReportRow]?
}

struct WorkerReportDetailData: Decodable {
    let id: String
    let status: String
    let managerNote: String?
    let taskId: String?
    let workerNote: String?

    enum CodingKeys: String, CodingKey {
        case id
        case status
        case managerNote = "manager_note"
        case taskId = "task_id"
        case workerNote = "worker_note"
    }
}

private struct WorkerReportDetailEnvelope: Decodable {
    let data: WorkerReportDetailData?
}

File: android/shared/src/main/java/ai/aistroyka/shared/WorkerApi.kt
package ai.aistroyka.shared

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

/**
 * Worker endpoints matching iOS [WorkerAPI] and `/api/v1` routes.
 */
object WorkerApi {
    private val uploadHttp = OkHttpClient.Builder()
        .connectTimeout(45, TimeUnit.SECONDS)
        .readTimeout(120, TimeUnit.SECONDS)
        .writeTimeout(120, TimeUnit.SECONDS)
        .build()

    suspend fun config(): ConfigResponse = ApiClient.request("config")

    suspend fun projects(): List<ProjectDto> {
        val r: ProjectsResponse = ApiClient.request("projects")
        return r.data.orEmpty()
    }

    /** POST /api/v1/devices/register — FCM token; use when wiring push (parity with iOS [WorkerAPI.registerDevice]). */
    suspend fun registerDevice(fcmToken: String) {
        val body = RegisterDeviceBody(
            deviceId = DeviceContext.deviceId,
            platform = "android",
            token = fcmToken.trim(),
        )
        val json = ApiClient.json.encodeToString(RegisterDeviceBody.serializer(), body)
        ApiClient.requestVoid(
            path = "devices/register",
            method = "POST",
            jsonBody = json,
            idempotencyKey = DeviceContext.idempotencyKeyDeviceRegister(fcmToken),
        )
    }

    /** POST /api/v1/devices/unregister — remove device_tokens row before sign-out (parity with iOS). */
    suspend fun unregisterDevice() {
        val body = UnregisterDeviceBody(deviceId = DeviceContext.deviceId)
        val json = ApiClient.json.encodeToString(UnregisterDeviceBody.serializer(), body)
        ApiClient.requestVoid(
            path = "devices/unregister",
            method = "POST",
            jsonBody = json,
            idempotencyKey = DeviceContext.idempotencyKeyDeviceUnregister(),
        )
    }

    suspend fun tasksToday(projectId: String?): List<TaskDto> {
        val q = if (!projectId.isNullOrBlank()) "?project_id=${java.net.URLEncoder.encode(projectId, Charsets.UTF_8.name())}" else ""
        val r: TasksTodayResponse = ApiClient.request("worker/tasks/today$q")
        return r.data.orEmpty()
    }

    suspend fun startDay(idempotencyKey: String): String {
        val r: WorkerDayResponse = ApiClient.request(
            path = "worker/day/start",
            method = "POST",
            jsonBody = "{}",
            idempotencyKey = idempotencyKey,
        )
        return r.data?.id ?: throw ApiError(null, null, "No worker day id in response")
    }

    suspend fun endDay(idempotencyKey: String) {
        ApiClient.requestVoid(
            path = "worker/day/end",
            method = "POST",
            jsonBody = "{}",
            idempotencyKey = idempotencyKey,
        )
    }

    /** GET /api/v1/tasks/:id — detail when assigned to current worker (or manager). Uses [TaskDetailResponse] from ManagerDtos. */
    suspend fun task(taskId: String): TaskDetailDto {
        val r: TaskDetailResponse = ApiClient.request("tasks/${taskId.trim()}")
        return r.data ?: throw ApiError(null, null, "No task in response")
    }

    suspend fun createReport(dayId: String?, taskId: String?, idempotencyKey: String): String {
        val body = ReportCreateBody(dayId = dayId, taskId = taskId)
        val json = ApiClient.json.encodeToString(ReportCreateBody.serializer(), body)
        val r: ReportCreateResponse = ApiClient.request(
            path = "worker/report/create",
            method = "POST",
            jsonBody = json,
            idempotencyKey = idempotencyKey
        )
        val id = r.data?.id ?: throw ApiError(null, null, "No report id in response")
        return id
    }

    suspend fun createUploadSession(purpose: String, idempotencyKey: String): Pair<String, String> {
        val body = CreateUploadSessionBody(purpose = purpose)
        val json = ApiClient.json.encodeToString(CreateUploadSessionBody.serializer(), body)
        val r: UploadSessionResponse = ApiClient.request(
            path = "media/upload-sessions",
            method = "POST",
            jsonBody = json,
            idempotencyKey = idempotencyKey
        )
        val data = r.data ?: throw ApiError(null, null, "No session data")
        val path = data.uploadPath ?: "media/${data.id}"
        return data.id to path
    }

    suspend fun finalizeUploadSession(
        sessionId: String,
        objectPath: String,
        mimeType: String?,
        sizeBytes: Int?,
        idempotencyKey: String,
    ) {
        val body = FinalizeUploadBody(
            objectPath = objectPath,
            mimeType = mimeType,
            sizeBytes = sizeBytes
        )
        val json = ApiClient.json.encodeToString(FinalizeUploadBody.serializer(), body)
        ApiClient.requestVoid(
            path = "media/upload-sessions/$sessionId/finalize",
            method = "POST",
            jsonBody = json,
            idempotencyKey = idempotencyKey
        )
    }

    suspend fun addMedia(reportId: String, uploadSessionId: String, idempotencyKey: String) {
        val body = AddMediaBody(reportId = reportId, uploadSessionId = uploadSessionId)
        val json = ApiClient.json.encodeToString(AddMediaBody.serializer(), body)
        ApiClient.requestVoid(
            path = "worker/report/add-media",
            method = "POST",
            jsonBody = json,
            idempotencyKey = idempotencyKey
        )
    }

    suspend fun submitReport(reportId: String, taskId: String?, idempotencyKey: String, workerNote: String? = null) {
        val json = buildSubmitReportJson(reportId, taskId, workerNote)
        ApiClient.requestVoid(
            path = "worker/report/submit",
            method = "POST",
            jsonBody = json,
            idempotencyKey = idempotencyKey
        )
    }

    /** GET /api/v1/worker/sync — recent reports for signed-in worker (status feed incl. changes_requested/approved). */
    suspend fun workerSync(): List<WorkerSyncReportRow> {
        val env: WorkerSyncEnvelope = ApiClient.request("worker/sync")
        return env.data?.reports.orEmpty()
    }

    /** GET /api/v1/reports/:id — own report detail with manager note and media metadata. */
    suspend fun reportDetail(id: String): WorkerReportDetailData {
        val env: WorkerReportDetailEnvelope = ApiClient.request("reports/${id.trim()}")
        return env.data ?: throw ApiError(null, null, "No report data")
    }

    suspend fun syncBootstrap(): SyncBootstrapResponse {
        return ApiClient.request("sync/bootstrap")
    }

    /**
     * On 409 returns [SyncConflictError] with mustBootstrap/serverCursor.
     */
    suspend fun syncChanges(cursor: Int, limit: Int = 100): SyncChangesResponse {
        val (body, code) = ApiClient.requestDataAndStatus("sync/changes?cursor=$cursor&limit=$limit")
        if (code == 409) {
            val conflict = try {
                ApiClient.json.decodeFromString(SyncConflictBody.serializer(), body)
            } catch (_: Exception) {
                throw ApiError(code, null, body.ifBlank { "Sync conflict" })
            }
            throw SyncConflictError(conflict)
        }
        if (code >= 400) {
            throw ApiError.fromHttp(code, body)
        }
        return ApiClient.json.decodeFromString(SyncChangesResponse.serializer(), body)
    }

    suspend fun syncAck(cursor: Int, idempotencyKey: String) {
        val body = """{"cursor":$cursor}"""
        ApiClient.requestVoid(
            path = "sync/ack",
            method = "POST",
            jsonBody = body,
            idempotencyKey = idempotencyKey
        )
    }

    /**
     * POST binary to Supabase Storage `media` bucket (same as iOS [UploadManager.uploadToSupabaseStorage]).
     * [pathInBucket] is the object key inside the bucket (no `media/` prefix).
     */
    suspend fun uploadToSupabaseStorage(data: ByteArray, pathInBucket: String) = withContext(Dispatchers.IO) {
        val base = AppRuntime.supabaseUrl.trimEnd('/')
        if (base.isBlank()) throw ApiError(null, null, "Supabase URL not configured")
        val token = SessionStore.getAccessToken()
            ?: throw ApiError(null, null, "Not signed in")
        val url = "$base/storage/v1/object/media/$pathInBucket"
        val req = Request.Builder()
            .url(url)
            .post(data.toRequestBody("image/jpeg".toMediaType()))
            .header("Authorization", "Bearer $token")
            .header("apikey", AppRuntime.supabaseAnonKey)
            .header("Content-Type", "image/jpeg")
            .build()
        val res = uploadHttp.newCall(req).execute()
        if (!res.isSuccessful) {
            val msg = if (res.code == 401 || res.code == 403) {
                "Storage policy denied. Check Supabase RLS for bucket media and tenant path."
            } else {
                "Storage upload failed (HTTP ${res.code})"
            }
            throw ApiError(res.code, null, msg)
        }
    }

    @Serializable
    private data class UnregisterDeviceBody(@SerialName("device_id") val deviceId: String)

    @Serializable
    private data class RegisterDeviceBody(
        @SerialName("device_id") val deviceId: String,
        val platform: String,
        val token: String,
    )

    @Serializable
    private data class ReportCreateBody(
        @SerialName("day_id") val dayId: String? = null,
        @SerialName("task_id") val taskId: String? = null,
    )

    @Serializable
    private data class CreateUploadSessionBody(val purpose: String)

    @Serializable
    private data class FinalizeUploadBody(
        @SerialName("object_path") val objectPath: String,
        @SerialName("mime_type") val mimeType: String? = null,
        @SerialName("size_bytes") val sizeBytes: Int? = null,
    )

    @Serializable
    private data class AddMediaBody(
        @SerialName("report_id") val reportId: String,
        @SerialName("upload_session_id") val uploadSessionId: String,
    )

}
